subsys/net/lib/http_server/src/server.c

```diff
@@ -0,0 +1,538 @@
+
+    if (clientIndex == ctx->num_clients) {
+	while (ctx->num_clients > 0 &&
+	       ctx->client_fds[ctx->num_clients].fd == 0) {
+	    --ctx->num_clients;
+	}
+    }
+}
+
+int accept_new_client(int server_fd)
+{
+    int new_socket;
+    struct sockaddr_in address;
+    int addrlen = sizeof(address);
+
+    new_socket =
+	accept(server_fd, (struct sockaddr *)&address, (socklen_t *)&addrlen);
+    if (new_socket < 0) {
+	perror("accept failed");
+	return -1;
+    }
+
+    return new_socket;
+}
+
+void handle_http1_request(struct http2_server_ctx *ctx, int i)
+{
+    const char *data;
+    size_t len;
+    data = content;
+    len = sizeof(content);
+
+    http_parser_init(&parser, HTTP_REQUEST);
+    http_parser_settings_init(&parserSettings);
+    parserSettings.on_url = on_url;
+
+    http_parser_execute(&parser, &parserSettings, buffer, BUFFER_SIZE);
+
+    if (strcmp(url_buffer, "/") == 0) {
+	sprintf(http_response,
+		"HTTP/1.1 200 OK\r\n"
+		"Content-Type: text/html\r\n"
+		"Content-Encoding: gzip\r\n"
+		"Content-Length: %u\r\n\r\n",
+		len);
+	if (sendall(ctx->client_fds[i].fd, http_response,
+		    strlen(http_response)) < 0) {
+	    close_client_connection(ctx, i);
+	} else {
+
+	    if (sendall(ctx->client_fds[i].fd, data, len) < 0) {
+		perror("sendall failed");
+		close_client_connection(ctx, i);
+	    }
+	}
+    } else {
+	const char *not_found_response = "HTTP/1.1 404 Not Found\r\n"
+					 "Content-Length: 9\r\n\r\n"
+					 "Not Found";
+	if (sendall(ctx->client_fds[i].fd, not_found_response,
+		    strlen(not_found_response)) < 0) {
+	    close_client_connection(ctx, i);
+	}
+    }
+    close_client_connection(ctx, i);
+}
+
+void handle_http2_request(struct http2_server_ctx *ctx, int i, int valread)
+{
+    printf("Hello HTTP/2.\n");
+    unsigned char frame[100];
+    ssize_t readBytes;
+    unsigned int frame_count;
+    int stream_header_id = 1;
+
+    if (hasUpgradeHeader == 0) {
+	const char *response = "HTTP/1.1 101 Switching Protocols\r\n"
+			       "Connection: Upgrade\r\n"
+			       "Upgrade: h2c\r\n"
+			       "\r\n";
+	if (sendall(ctx->client_fds[i].fd, response, strlen(response)) < 0) {
+	    close_client_connection(ctx, i);
+	}
+
+	/* Read the client data */
+	valread = recv(ctx->client_fds[i].fd, buffer, BUFFER_SIZE, 0);
+	printf("valread length: %u\n", valread);
+	if (valread < 0) {
+	    perror("read failed");
+	    close_client_connection(ctx, i);
+	} else if (valread == 0) {
+	    printf("Connection closed by peer.\n");
+	    close_client_connection(ctx, i);
+	} else {
+	    printf("buffer %s\n", buffer);
+
+	    /* Check the client preface */
+	    if (strncmp(buffer, "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n", 24) != 0) {
+		printf("The client does not support HTTP/2.\n");
+	    } else {
+		printf("The client support HTTP/2.\n");
+	    }
+	}
+    }
+
+    if (valread > 24) {
+	/* Define a new buffer for the remaining data and copy it there */
+	unsigned char ubuffer[BUFFER_SIZE];
```

Comment:
We should really avoid declaring large buffers on a stack. I think we should have one buffer statically allocated per client, and use only it for communication.

Reply:
And actually it should be even needed to copy the buffers, just offset the orignal buffer when providing the pointer to the function (so you don't provide `buffer` directly, but `buffer + some_offset`).